src/audio/rtnr/rtnr.c

```diff
@@ -0,0 +1,554 @@
+	#endif // RTNR_ENABLE_PROCESSING
+
+	/* Done. */
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+err:
+	rfree(dev);
+	rfree(cd);
+	return NULL;
+}
+
+static void rtnr_free(struct comp_dev *dev)
+{
+	struct comp_data *cd = comp_get_drvdata(dev);
+
+	comp_info(dev, "rtnr_free()");
+
+	#if RTNR_ENABLE_PROCESSING
+	RTKMA_API_Context_Free((void *)&cd->rtk_agl);
+	free_RtkMemoryPool();
+	#endif // RTNR_ENABLE_PROCESSING
+
+	rfree(cd);
+	rfree(dev);
+}
+
+/* set component audio stream parameters */
+static int rtnr_params(struct comp_dev *dev, struct sof_ipc_stream_params *params)
+{
+	int ret;
+
+	comp_info(dev, "rtnr_params()");
+
+	ret = comp_verify_params(dev, 0, params);
+	if (ret < 0) {
+		comp_err(dev, "rtnr_params() error: comp_verify_params() failed.");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int rtnr_cmd_get_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata, int max_size)
+{
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_get_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+static int rtnr_cmd_set_data(struct comp_dev *dev,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	int ret = 0;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		comp_info(dev, "rtnr_cmd_set_data(), SOF_CTRL_CMD_BINARY");
+		break;
+	default:
+		comp_err(dev, "rtnr_cmd_set_data() error: invalid command %d", cdata->cmd);
+		ret = -EINVAL;
+		break;
+	}
+	return ret;
+}
+
+/* used to pass standard and bespoke commands (with data) to component */
+static int rtnr_cmd(struct comp_dev *dev, int cmd, void *data,
+		   int max_data_size)
+{
+	struct sof_ipc_ctrl_data *cdata = data;
+	int ret = 0;
+
+	comp_info(dev, "rtnr_cmd()");
+
+	switch (cmd) {
+	case COMP_CMD_SET_DATA:
+		ret = rtnr_cmd_set_data(dev, cdata);
+		break;
+	case COMP_CMD_GET_DATA:
+		ret = rtnr_cmd_get_data(dev, cdata, max_data_size);
+		break;
+	}
+
+	return ret;
+}
+
+static int rtnr_trigger(struct comp_dev *dev, int cmd)
+{
+	comp_info(dev, "rtnr_trigger() cmd: %d", cmd);
+
+	return comp_set_state(dev, cmd);
+}
+
+/* copy and process stream data from source to sink buffers */
+static int rtnr_copy(struct comp_dev *dev)
+{
+	struct list_item *blist;
+	struct comp_buffer *sink;
+	struct comp_buffer *source;
+	int frames;
+	int num_blocks;
+	int sink_bytes;
+	int source_bytes;
+	const struct audio_stream *sources_stream[RTNR_MAX_SOURCES] = { NULL };
+	struct comp_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *sources[RTNR_MAX_SOURCES] = { NULL };
+	uint32_t flags = 0;
+	int num_sources = 0;
+
+	comp_dbg(dev, "rtnr_copy()");
+
+	/* align source streams with their respective configurations */
+	list_for_item(blist, &dev->bsource_list) {
+		source = container_of(blist, struct comp_buffer, sink_list);
+		buffer_lock(source, &flags);
+		if (source->source->state == dev->state) {
```

Comment:
are you expecting to behave like a mixer? Why not let the mixer do the mixing? Or are you trying to route all paths through a single instantiation, which is not the design behind SOF, there should be separate state instantiations per path so you do not block your dmic or headphone mic path if the other is being used